### PR TITLE
Fix handling of lazy vectors in FilterProject operator

### PR DESCRIPTION
**Pull request by bikramSingh91**
Currently the expression evaluation via ExprSet keeps track of all
input fields that are multiply referenced among all expressions
and pre-loads them before starting evaluation. However, a
FilterProject operator can have a field that is singly referenced
among its filter and projection expressions but is also an identity
projection. This patch makes sure that these kind of fields are
identified and pre-loaded to avoid running into issues of partially
loaded lazy vectors.

Test Plan:
Added unit test

### CODE REVIEW

**Comment by bikramSingh91** [comment]
related to  #2501

File: velox/exec/FilterProject.h [hunk 1]
```diff
@@ -81,5 +81,15 @@ class FilterProject : public Operator {
...
+
+  // Indices for fields/input columns that are both an identity projection and
+  // are referenced by either a filter or project expression. This is used to
+  // identify fields that need to be preloaded before evaluating filters or
+  // projections.
+  // Consider projection with 2 expressions: f(c0) AND g(c1), c1
+  // If c1 is a LazyVector and f(c0) AND g(c1) expression is evaluated first, it
+  // will load c1 only for rows where f(c0) is true. However, c1 identity
+  // projection needs all rows.
+  std::vector<int32_t> multiplyReferencedFieldIndices_;
```
Comment (mbasmanova) [inline velox/exec/FilterProject.h:93]: column_index_t

File: velox/exec/FilterProject.cpp [hunk 2]
```diff
@@ -75,6 +76,22 @@ FilterProject::FilterProject(
   }
   numExprs_ = allExprs.size();
   exprs_ = makeExprSetFromFlag(std::move(allExprs), operatorCtx_->execCtx());
+
+  if (numExprs_ > 0 && !identityProjections_.empty()) {
+    auto inputType = project ? project->sources()[0]->outputType()
+                             : filter->sources()[0]->outputType();
+    std::unordered_set<uint32_t> distinctFieldIndexes;
```
Comment (mbasmanova) [inline velox/exec/FilterProject.cpp:83]: nit: distinctFieldIndexes -> xxxIndices

File: velox/exec/FilterProject.cpp [hunk 3]
```diff
@@ -117,6 +134,13 @@ RowVectorPtr FilterProject::getOutput() {
   auto* rows = localRows.get();
   rows->setAll();
   EvalCtx evalCtx(operatorCtx_->execCtx(), exprs_.get(), input_.get());
+
+  // Pre-load lazy vectors which are referenced by both expressions and identity
```
Comment (mbasmanova) [inline velox/exec/FilterProject.cpp:138]: If someone is using expression evaluation in a standalone manner, they would need to do this as well. Should we maybe document that standalone evaluation is not supported for Lazy Vectors?

**Review by mbasmanova — APPROVED** [review]
@bikramSingh91 Thank you investigating and fixing this issue and doing so very quickly. I love the way you write comments.

File: velox/exec/FilterProject.cpp (hunk 3 above)
Comment (bikramSingh91) [reply velox/exec/FilterProject.cpp:138]: Good idea! How about we recommend both the pro option (figure out which lazy vectors are referenced within expressions and outside and load them, see FilterProject operator as reference) and the easy/cautious option (load all lazy vectors before running expr eval)

Comment (mbasmanova) [reply velox/exec/FilterProject.cpp:138]: Sounds great.

File: velox/exec/FilterProject.cpp (hunk 2 above)
Comment (bikramSingh91) [reply velox/exec/FilterProject.cpp]: done

File: velox/exec/FilterProject.cpp (hunk 3 above)
Comment (bikramSingh91) [reply velox/exec/FilterProject.cpp:139]: done

File: velox/exec/FilterProject.h (hunk 1 above)
Comment (bikramSingh91) [reply velox/exec/FilterProject.h]: done

**Pull request closed without merging**